consensus/safety-rules/src/safety_rules.rs

```diff
@@ -134,10 +134,14 @@ impl SafetyRules {
     /// committed block, might panic otherwise.
     /// The update function is invoked whenever a system learns about a potentially high QC.
     pub fn update(&mut self, qc: &QuorumCert) {
-        // Preferred block rule: choose the highest 2-chain head.
         if qc.parent_block().round() > self.persistent_storage.preferred_round() {
             self.persistent_storage
                 .set_preferred_round(qc.parent_block().round());
+        } else if qc.ledger_info().ledger_info().epoch() > self.persistent_storage.epoch() {
```

Comment:
this probably should happen before we check the parent_block().round() otherwise we don't update the preferred_round

I'd prefer to have a separate API to notify epoch changes so it's clear `update` is with-in the same epoch.

Reply:
fair point on the first one

on the second point, that's fair is that in scope for this PR?

Reply:
sure, not for this PR, it's already too big

Reply:
done